src/sdk/round.rs

```diff
@@ -199,30 +199,87 @@ impl<F, K, P, const MAX_MSG_IN_LEN: usize> Round<F, K, P, MAX_MSG_IN_LEN> {
+    /// Why? We want to allow robust protocols to succeed even when faulters are present.
     pub fn execute_next_round(mut self) -> TofnResult<Protocol<F, K, P, MAX_MSG_IN_LEN>> {
         let my_share_id = self.info().share_info().my_id();
         let my_party_id = self.info().party_id();
         let curr_round_num = self.info.round();
 
-        if !self.msg_in_faulters.is_empty() {
-            warn!(
-                "peer {} (party {}) says: faulters detected during msg_in: ending protocol in round {}",
-                my_share_id, my_party_id, curr_round_num,
-            );
+        self.info.advance_round();
 
-            return Ok(Protocol::Done(Err(self.msg_in_faulters)));
-        }
+        // delete any messages received from msg_in faulters before executing the round
+        let faulter_share_ids = self
+            .info
+            .party_share_counts()
+            .share_id_subset(&self.msg_in_faulters.as_subset()?)?;
 
-        self.info.advance_round();
+        for faulter_share_id in faulter_share_ids {
+            self.bcasts_in.unset(faulter_share_id)?;
+            self.p2ps_in.unset_all(faulter_share_id)?;
+        }
 
-        self.round
+        // execute the round
+        let protocol_status = self
+            .round
             .execute_raw(
                 self.info.share_info(),
                 self.bcasts_in,
                 self.p2ps_in,
                 self.expected_msg_types,
             )?
-            .build(self.info)
+            .build(self.info)?;
+
+        // if no faulters then we're done
+        if self.msg_in_faulters.is_empty() {
+            return Ok(protocol_status);
+        }
+
+        // msg_in_faulters found; decide how to report them
+        match protocol_status {
+            Protocol::NotDone(ref round) => {
+                warn!(
+                    "peer {} (party {}) says: tofn SDK detected faulters in msg_in round {} but protocol still in progress at round {}",
+                    my_share_id, my_party_id, curr_round_num, round.info.round(),
+                );
+                Ok(protocol_status)
+            }
+            Protocol::Done(Ok(_)) => {
+                if !self.msg_in_faulters.is_empty() {
+                    warn!(
+                        "peer {} (party {}) says: tofn SDK detected faulters in msg_in round {} but protocol ended in happy path",
+                        my_share_id, my_party_id, curr_round_num,
+                    );
+                }
+                Ok(protocol_status)
+            }
+            Protocol::Done(Err(ref faulters)) => {
+                info!(
+                    "peer {} (party {}) says: tofn SDK detected faulters in msg_in round {} and protocol ended in sad path; merging faulters lists",
+                    my_share_id, my_party_id, curr_round_num,
+                );
+
+                // merge faulters into msg_in_faulters and return msg_in_faulters
```

Comment:
> What I had in mind was to convert the party faulters set to share faulters set by just faulting all shares of that party and passing that as an argument to execute_raw. This works since you're already unsetting messages for all shares of the party.

True, but we already achieve this end by unsetting all the messages.  (Recall that a missing message always results in that share being marked as a faulter.)

If we were to go this route then it might make more sense _not_ to unset all the messages from that party.  Instead, let the protocol decide whether to accept messages from a party known to be a faulter.

I like this idea because it leaves the protocol implementation with more control over whom to accuse.

> Now that I think about it more, the decision to unset messages from all shares makes sense even if some shares weren't faulting, since otherwise it would lead to the weird scenario in a robust protocol were it ended in the happy path, along with one faulter (one of whose share was faulting) but another non-faulting share was still used to obtain the protocol output.

I don't think there's anything wrong with that scenario.  One can imagine an execution that ends in happy path with an adversary who corrupts only one of his shares and yet if we disqualify all the shares from that adversary then the protocol no longer has sufficient honest shares to end in happy path.  Effectively, our policy would turn a happy path execution into a sad path execution, which is contrary to the intent of a robust protocol.

Besides, even with the current policy of deleting all messages from a `msg_in` faulter, it's still possible for an adversary to create the above outcome simply by behaving well enough to satisfy `msg_in` and then corrupting only a single one of his shares.

Reply:
> If we were to go this route then it might make more sense not to unset all the messages from that party. Instead, let the protocol decide whether to accept messages from a party known to be a faulter.

I added a commit for this.  It simplified `execute_next_round` quite a bit.